js/main.js

```diff
@@ -61,78 +61,80 @@ if (ACTIVE_ENV !== '') {
   use_envs = JSON.parse(localStorage.getItem('envIDs')) || ['main'];
 }
 
-// TODO: Move some of this to smaller components and/or use something like redux
-// to move state out of the app to a standalone store.
-class App extends React.Component {
-  state = {
-    connected: false,
-    readonly: false,
-    sessionID: null,
+function App() {
+  // -------------- //
+  // state varibles //
+  // -------------- //
+
+  // internal variables
+  const mounted = useRef(false);
+  const [connected, setConnected] = useState(false);
+  const [sessionInfo, setSessionInfo] = useState({ id: null, readonly: false });
+  const [resizeClickHappened, setResizeClickHappened] = useState(false);
+  const windowSize = useRef({
+    // width: 1280,
+    // cols: 100,
+  });
+
+  // data stores
+  const [storeMeta, setStoreMeta] = useState({
+    envList: ENV_LIST.slice(),
+    layoutLists: new Map([['main', new Map([[DEFAULT_LAYOUT, new Map()]])]]),
+  });
+  const [storeData, setStoreData] = useState({
     panes: {},
-    consistent_pane_copy: {},
-    focusedPaneID: null,
+    layout: [],
+  });
+
+  // user-changeable
+  const [showEnvModal, setShowEnvModal] = useState(false);
+  const [showViewModal, setShowViewModal] = useState(false);
+  const [focusedPaneID, setFocusedPaneID] = useState(null);
+  const [selection, setSelection] = useState({
     envID: use_env,
     envIDs: use_envs,
     layoutID: DEFAULT_LAYOUT,
-    // Bad form... make a copy of the global var we generated in python.
```

Comment:
I still think this isn't great form, but alas. Not to resolve here, but maintain the note

Reply:
Adapted & amended first commit. (I found it cleaner that way)